framework/src/functormaterials/FunctorSmoother.C

```diff
@@ -0,0 +1,219 @@
+    rm_params.template set<std::string>("for_whom") = name();
+    rm_params.template set<MooseMesh *>("mesh") = &const_cast<MooseMesh &>(_mesh);
+    rm_params.template set<Moose::RelationshipManagerType>("rm_type") =
+        Moose::RelationshipManagerType::GEOMETRIC | Moose::RelationshipManagerType::ALGEBRAIC |
+        Moose::RelationshipManagerType::COUPLING;
+    rm_params.template set<unsigned short>("layers") = 1;
+    rm_params.template set<bool>("use_point_neighbors") = false;
+    rm_params.template set<bool>("attach_geometric_early") = false;
+    rm_params.template set<bool>("use_displaced_mesh") =
+        parameters.template get<bool>("use_displaced_mesh");
+    mooseAssert(rm_params.areAllRequiredParamsValid(),
+                "All relationship manager parameters should be valid.");
+    auto rm_obj = factory.template create<RelationshipManager>(
+        "ElementSideNeighborLayers", name() + "_functor_smoothing", rm_params);
+
+    // Delete the resources created on behalf of the RM if it ends up not being added to the
+    // App.
+    if (!_app.addRelationshipManager(rm_obj))
+      factory.releaseSharedObjects(*rm_obj);
+  }
+
+  for (const auto i : index_range(_functors_in))
+  {
+    // This potentially upcasts functors from non-AD to AD
+    const auto & functor_in = getFunctor<typename Moose::ADType<T>::type>(_functors_in[i]);
+
+    // We always add the AD type
+    addFunctorProperty<typename Moose::ADType<T>::type>(
+        _functors_out[i],
+        [this, &functor_in](const auto & r, const auto & t) -> typename Moose::ADType<T>::type
+        {
+          typename Moose::ADType<T>::type base_value = functor_in(r, t);
+          typename Moose::ADType<T>::type average = 0;
+
+          const Elem * r_elem = nullptr;
+          if constexpr (std::is_same_v<const Moose::ElemArg &, decltype(r)>)
+            r_elem = r.elem;
+
+          // Handle the single isolated element case
+          if (r_elem && r_elem->n_neighbors() == 0)
+            return base_value;
+
+          // If these techniques are of interest elsewhere, they should be moved to a
+          // FunctorAveragingUtils file and namespaced
+          if (_smoothing_technique == FACE_AVERAGE)
+          {
+            if constexpr (std::is_same_v<const Moose::ElemArg &, decltype(r)>)
+            {
+              unsigned int n_faces = 0;
+              for (const auto side_index : r_elem->side_index_range())
+              {
+                Moose::FaceArg face_arg{_mesh.faceInfo(r_elem, side_index),
+                                        Moose::FV::LimiterType::CentralDifference,
+                                        true,
+                                        false,
+                                        nullptr};
+                if (face_arg.fi)
+                {
+                  average += functor_in(face_arg, t);
+                  n_faces++;
+                }
+              }
+              average /= n_faces;
```

Comment:
I believe that this is exactly what is being done conceptually in `Reconstructions.h` for `interpolateReconstruct` which is what is used for the porosity smoothing. Perhaps this logic could be factored out although I wouldn't require it

Reply:
yes. Am I correct that interpolateReconstruct is meant to build the reconstructed functor once and then provide it as a map functor? whereas here it s more on-the-fly (unless you use caching)

Reply:
yes that's right